### ISSUE TITLE
How to deal with objects only partially in training images

### ISSUE THREAD

**Issue by s-ilic** [AlexeyAB/darknet#7934]
I am trying to train a YOLO (v3) network on a set of images, but I am faced with a problem when preparing the training set and the corresponding annotations.

Sometimes, an image contains my object of interest, but only half of the object is visible in the image (e.g. half a cat, with the rest out of frame). How do I deal with that?

Should I:
- Still including the bounding box of the partial object in the training set?
- Only include bounding boxes for objects which are fully in the frame (or whose center is in the frame maybe)?
- Only include bounding boxes of objects whose "fractional area in frame" is above a certain threshold (e.g. objects that are 90% in frame)?

And if I need to include partial objects, should I give the bounding box containing the part of the object visible in the frame? Or should I give the actual bounding box (which would thus go over the edges of the images, and thus may have e.g. negative coordinates) of the "true", full object (which I actually have access to)?

**Comment by stephanecharette** [comment]
See:  https://www.ccoderun.ca/darkmark/ImageMarkup.html

**Comment by s-ilic** [comment]
Thank you for the link, @stephanecharette ; however, I already went through it a while ago but didn't really find any concrete and quantitative answer to my questions...

For example, it says:
> Generally, if most of an object is visible, then it should be tagged.

So does that mean if more than 50 % of an object is in the image, it should be tagged ?

**Comment by stephanecharette** [comment]
Depends on your object and how distinctive it is, how large it is, if it can be confused with another similar object you're training for, etc.  Best thing to do is try it and see.

If mosaic, tiling, and crop+zoom has taught me anything recently, it is that these small details don't really matter as much as you may think.  Once you have enough marked up images, having a few where you wondered if you should or should not annotate a partial object probably won't impact your network.

**Comment by s-ilic** [comment]
I see...

I'm in a bit of a peculiar situation: essentially, I have this gigantic training image (think something like 100,000 x 100,000) which contains a certain number of objects (I know the bounding boxes of all of them), and they all belong to a single class.

Currently, I subdivide this gigantic image into smaller images (e.g. 1,000 x 1,000), and train my network on those. Incidentally, that's how I end up with the problem of objects that can be only partially in an image.

Of course, I could circumvent the problem by cutting the smaller images in a way that always fully contains the objects; however that's not realistic, because when I'll apply this network in a "real" setting, I obviously won't be able to do such cutting and will likely stick to some regular tiling of the big image...

**Comment by stephanecharette** [comment]
You should use DarkMark and DarkHelp's tiling feature:  https://www.ccoderun.ca/darkmark/darknet_images.html#tile_images

**Comment by s-ilic** [comment]
Thanks for the link, @stephanecharette .

I understand well the points they are making, and the guidelines they give when it comes to huge images is essentially what I am doing already; the issue is still that sometimes, well... my tiling cuts an object in two! And they don't explicitly say how to deal with those cases...

**Comment by stephanecharette** [comment]
But the good news is DarkHelp and DarkMark already handles this.  :)  You shouldn't ever have to worry about it.

The only thing you may need to tweak are the thresholds for how objects are put back together when they cross the boundary of tiles.  If the defaults don't work for you, then you may need to tweak these:

- [only_combine_similar_predictions](https://www.ccoderun.ca/darkhelp/api/classDarkHelp.html#aa119e80698e9dc874abed0cf9f60a125)
- [tile_edge_factor](https://www.ccoderun.ca/darkhelp/api/classDarkHelp.html#af0ba83fc66243e9a7559765d4afcf820)
- [tile_rect_factor](https://www.ccoderun.ca/darkhelp/api/classDarkHelp.html#ad57d399230133c8d20dbab5514782a1a)

But hopefully the defaults are decent and work well for you, in which case you simply let DarkHelp do the work of tiling.

**Comment by s-ilic** [comment]
Oh, I just realized that you are actually the author of DarkHelp and DarkMark, sorry! Very neat applications!

The code that I'm currently using is actually based on a pure Python/Tensorflow implementation of YoloV3, and the tiling/bounding boxes/annotations are done via a custom set of codes that I developed myself. As a consequence, it is not really straightforward for me to directly incorporate DarkHelp and DarkMark in my workflow...

If I understand correctly, the three functions you are referring to are related to the way one has to potentially combine bounding boxes that lie near the edges of neighboring tiles -- and thus may actually be the same object. But since you're very familiar with the code, do you have in mind how the bounding boxes are defined for an object that is "cut in two" by the tiling? If not, could you point me to the part of the source code that deals with this? Thanks in advance!

**Comment by stephanecharette** [comment]
This is the code that recombines predictions back together when they've been sliced due to large images + tiles:  https://github.com/stephanecharette/DarkHelp/blob/master/src-lib/DarkHelp.cpp#L336-L502

**Comment by s-ilic** [comment]
That's for predictions using an already trained network, correct?

My question was rather about what happens at training time: are "big" training images tiled as well? And how are the "ground truth" bounding boxes affected by the tiling?

**Comment by stephanecharette** [comment]
Yes, DarkMark supports tiling for training as well.  I'm guessing the line you want to see there is this one: https://github.com/stephanecharette/DarkMark/blob/master/src-darknet/DarknetImages.cpp#L303
